Allow user to pass various gluster volume options when  glusterfs volumes are dynamically provisioned.

humblec opened the pull request:
This patch add new storage class parameter called
`volumeoptions` which can be used to set various volume options.
for eg# if you want to enable encryption on volumes,
the values like `client.ssl on`, `server.ssl on`..etc can be passed
to `volumeoptions` parameter in storageclass.
    

Issue # https://github.com/kubernetes/kubernetes/issues/28454

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/44174)
<!-- Reviewable:end -->

rootfs commented on examples/persistent-volume-provisioning/README.md:
```diff
@@ -109,6 +109,7 @@ parameters:
   gidMin: "40000"
   gidMax: "50000"
   volumetype: "replicate:3"
+  volumeoptions: "encryption"
```
wouldn't it be a boolean?

humblec replied:
@rootfs my plan is to have a string with more options, for eg, "encryption, performance"..etc in this attribute. So that, internally we will develop a logic in heketi to set volume options based on these values.  So keeping it as meaningful input strings from the user instead of a boolean.

thockin commented:
@k8s-bot ok to test

msau42 commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -1002,5 +1010,18 @@ func parseClassParameters(params map[string]string, kubeClient clientset.Interfa
+			case "encryption":
+				cfg.volumeOptions = append(cfg.volumeOptions, v)
+			default:
+				return nil, fmt.Errorf("StorageClass for provisioner %q must have valid ( for eg, 'encrypption') volume option", glusterfsPluginName)
```
Should be "encryption"

humblec replied:
Corrected.

msau42 commented on examples/persistent-volume-provisioning/README.md:
```diff
@@ -134,6 +135,8 @@ For example:
   'Distribute volume':
     `volumetype: none`
 
+* `volumeoptions` : This option allows you to specify the gluster volume option which has to be set on the provisioned volume. The value string is an abstraction of the real options set on the volume. At present the only available value which can be set by this attribute is `encryption`, if this value has been set, all the required volume options ( for eg. `client.ssl, server.ssl`) for enabling e ...
```
@jsafrane has a proposal out to be able pass parameters from the PVC through the storage class to the provisioner.  Would it make more sense to have one parameter per option, instead of combining all the options into one parameter?  This could potentially make it easier to parse and validate each option, as more are added.  And it could be better for readability in general.

humblec replied:
@msau42  Thanks. At present GlusterFS support almost 200 volume options. So having it passed per parameter is not possible. Most of the time 2/3/4 options will be passed by the administrator. More or less, we are not doing any parsing in glusterfs plugin any more, rather it has been pushed to next layer( heketi). so I think it should be fine for now.

jeffvance commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -1009,5 +1016,18 @@ func parseClassParameters(params map[string]string, kubeClient clientset.Interfa
 		return nil, fmt.Errorf("StorageClass for provisioner %q must have gidMax value >= gidMin", glusterfsPluginName)
 	}
 
+	if len(parseVolumeOptions) != 0 {
```
Thinking ahead to when we (likely) move in-tree plugins out-of-tree, why should kubernetes be parsing these options? Would it be better to pass this []string as is to the provider?

humblec replied:
We are not doing parsing any more. The parsing logic was introduced when we want to do some kind of wrapping in the plugin. By wrapping I mean,  user only know about the string 'encryption', but internally we will have 'cleint.ssl on` , `server.ssl on` options tied to this string "encryption". I deviated from this path and we have agreed to move with direct passthrough of options. Thanks.

humblec commented:
/assign @jsafrane @rootfs

rootfs commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -966,6 +969,11 @@ func parseClassParameters(params map[string]string, kubeClient clientset.Interfa
 
+		case "volumeoptions":
+			if len(v) != 0 {
+				parseVolumeOptions = v
```
please validate the options

humblec replied:
@rootfs we are not validating any options from provisioner or in heketi. It has been passed to glusterd and validation will only happen on volume creation, because there are atleast 256 volume options available to set on gluster volume.  Validation in any of the above layers will be difficult for the same reason.

humblec replied:
@rootfs I have added the link to the GUIDE. Thanks.

rootfs replied:
This is unfortunate. Could there be a case a wrong option cause a service disruption? Volumes are created and space allocated but unusable due to unsupported options. If this happens, we need options printed in messages and logs.

humblec replied:
@rootfs no. If the option specified is wrong, the volume creation will fail and it will be logged in heketi as like any other volume creation failure from glusterd. Because options are tried as part of volume creation in heketi.

jsafrane commented on examples/persistent-volume-provisioning/README.md:
```diff
@@ -170,6 +170,9 @@ For example:
 
 For available volume types and its administration options refer: ([Administration Guide](https://access.redhat.com/documentation/en-US/Red_Hat_Storage/3.1/html/Administration_Guide/part-Overview.html))
 
+* `volumeoptions` : This option allows you to specify the gluster volume option which has to be set on the provisioned volume. The value string should be comma seperated strings which need to be set on the volume. As shown in example, if you want to enable encryption on gluster dynamically provisioned volumes you can pass `client.ssl on, server.ssl on` options. This is an optional parameter.
```
A link to gluster docs with list of parameters would be nice

humblec replied:
@jsafrane Thanks .. Addressed this comment.

humblec commented:
/test pull-kubernetes-federation-e2e-gce

humblec committed:
This patch add new storage class parameter called
`volumeoptions` which can be used to set various volume options.
for eg# if you want to enable encryption on volumes,
the values like `client.ssl on`, `server.ssl on`..etc can be passed
to `volumeoptions` parameter in storageclass.

Issue # https://github.com/kubernetes/kubernetes/issues/28454

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

ahmetb commented on examples/persistent-volume-provisioning/glusterfs/glusterfs-storageclass.yaml:
```diff
@@ -12,3 +12,4 @@ parameters:
   gidMin: "40000"
   gidMax: "50000"
   volumetype: "replicate:3"
+  volumeoptions: "client.ssl on, server.ssl on"
```
please undo this and submit it to https://github.com/kubernetes/examples

humblec replied:
Done.

humblec commented:
@k8s-bot kops aws e2e test this

rootfs commented:
/lgtm

jsafrane commented:
/release-note-none

humblec commented:
@rootfs @jsafrane Thanks ! Could you please bring back LGTM ? no changes in the code, I just made storageclass example file intact.

jsafrane commented:
/lgtm
